canonical/ubuntu-pro-client
# migrate user config

This is the PR that moves apt_news and similar config items (those that are intended to be managed via the `pro config` cli) out of the conffile.

This doesn't go so far as to rm_conffile uaclient.conf as some users may still expect it to be there.

There are several little changes in separate commits that are prereqs for the main migration commit.

See the comment in preinst and the migrate_user_config.py script for some more details.

Then there are separate commits with a test script to be used at SRU time of this migration, and of docs that will contain what to do if one of the unlikely errors occurs.

Also notice that we maintain backwards compatibility so even if migrations fail or if a user manually edits uaclient.conf later, we'll still respect `ua_config` settings in uaclient.conf. This is required because at least one other tool relies on this (cloud-init :eyes:)

All feedback welcome.

LP #2004280


To Do:
- [x] support old ua_config at runtime to support https://cloudinit.readthedocs.io/en/latest/reference/modules.html#ubuntu-advantage
    - medium term: change how this works in cloud-init (and optionally remove our support?)
    - medium term: add integration tests for all cloud-init pro options - SC-1427
- [x] fix remaining CI failures
- [x] search for `ua_config` and fix everywhere we use it to the new solution
- [x] add LP reference to main commit
- [x] add explanation doc with the contents of possible error messages and what to do about them
- [x] add additional unit tests for any of the small PRs?
- [x] additional testing of failure cases
- [x] add test script for SRU verification
- [x] integration test for still supporting ua_config in uaclient.conf
- [x] update version guards to 27.14~

## Test Steps


Try all kinds of upgrade scenarios you can think of: try various edits to uaclient.conf and then upgrade to this version. Does the UX make sense? Do conffile prompts only happen in reasonable scenarios?

Maybe also hack preinst to add failures in the python script and see what happens - is it good enough?

## Checklist


## Does this PR require extra reviews?
